numba/tests/test_entrypoints.py

```diff
@@ -0,0 +1,51 @@
+import sys
+import types
+
+import pkg_resources
+
+from .support import TestCase
+
+
+class TestEntrypoints(TestCase):
+    """
+    Test registration of init() functions from Numba extensions
+    """
+
+    def test_init_entrypoint(self):
+        # loosely based on Pandas test from:
+        #   https://github.com/pandas-dev/pandas/pull/27488
+
+        # FIXME: Python 2 workaround because nonlocal doesn't exist
+        counters = {'init': 0}
+
+        def init_function():
+            counters['init'] += 1
+
+        mod = types.ModuleType("_test_numba_extension")
+        mod.init_func = init_function
+        sys.modules[mod.__name__] = mod
+
+        # We are registering an entry point using the "numba" package
+        # ("distribution" in pkg_resources-speak) itself, though these are
+        # normally registered by other packages.
+        dist = "numba"
+        entrypoints = pkg_resources.get_entry_map(dist)
+        my_entrypoint = pkg_resources.EntryPoint(
+            "init", # name of entry point
+            mod.__name__, # module with entry point object
+            attrs=['init_func'], # name of entry point object
+            dist=pkg_resources.get_distribution(dist)
+        )
+        entrypoints.setdefault('numba_extensions', {})['init'] = my_entrypoint
+
+        import numba.entrypoints
+        numba.entrypoints._already_initialized = False  # Allow reinitialization
```

Comment:
I think this is mutating global state in a test and so the test should perhaps a) be serialized b) ensure that the state is not propagated outside of this test.

Reply:
Does it matter if we serialize this since the state change is confined to a worker process?  No changes on disk will happen, and workers don't run tests in parallel threads.

This does need a try-finally to ensure restoration of state.

Reply:
ok, added try-finally